app/models/concerns/host_searching.rb

```diff
@@ -109,9 +111,12 @@ def filter_by_compliance(_filter, operator, value)
     end
 
     def filter_by_compliance_score(_filter, operator, score)
-      ids = ::Host.includes(:test_result_profiles).select do |host|
-        host.compliance_score.public_send(operator, score.to_f)
+      unless NUM_OPERATORS.include?(operator)
+        raise ActiveRecord::StatementInvalid
       end
+
+      ids = ::Host.joins(:test_result_profiles)
+                  .where("test_results.score #{operator} ?", score.to_f)
       ids = ids.pluck(:id).map { |id| "'#{id}'" }
```

Comment:
Can it use the previous query as a subquery into the condition?
We are putting trust on the ids coming from Inventory.

Reply:
No, because we don't have access to the right context (profile/policy). Also if you read it further, we're building a "filter" based on those IDs and then it's being called on an already reduced subset of the Hosts in the context of the given profile/policy AND the user.

Reply:
Can you please elaborate on the profile/policy context that it makes the set reduced? 
The lookup by score here does not seem to be affected by any context. It would take hosts by their score from any of its test results, regardless of a policy.

Regarding the ids formatted to raw SQL, there is a `parameter` that can be passed within the returning hash with those ids. It could work, unless some other condition/query would shift the order of placeholders.

Reply:
There's a base query initiated by the browser and it already provides a reduced set of hosts (in our case under a policy). This helper query above gets a list of all IDs that meet the given conditions (in our case the score is below/above something) and the resulting IDs are being used to do a final sweep on the base query's result. The final result is an intersection of these two and this implementation was not much different before I introduced the changes.

Unfortunately there's no way to link these queries together as the second one is being called from inside `scoped_search` and the implementation doesn't allow us to pass any kind of context other than the three sanitized arguments.

Reply:
(as discussed on chat) this is a more complex problem. 
Imagine a scenario with a Host has two policies. Policy A in which it has score 20, and Policy B in which it has score 100. Query of `policy = A and compliance_score > 50` would still yield that host.

I blame the introduction of virtual fields that are context based.

Reply:
A possible solution could be to turn this around and instead of requesting hosts with (a limited set of) profiles underneath we could request the (limited set of) profiles and the hosts underneath. By turning the whole thing around, it should be possible to safely narrow down the results with my method in this PR.

Reply:
Nevermind, this is not gonna work, for the same reason :disappointed: `scoped_search` still won't see the parent

Reply:
So I guess we should go around `scoped_search` by introducing the `minScore` and `maxScore` optional arguments on the `testResultProfiles` and use that to narrow down our results. @akofink wdyt?

Reply:
After looking into this, I now see what you mean about the scoped search context where each search term gets a filter method. I agree that we're hitting the limits of scoped search and maybe should opt to implement this directly by introducing one or two new parameters. We can validate parameters exist when others are present (i.e. `policy_id` and `policy_score`) or do something weird like `policy_score[id]=8b409091-567f-47a7-b79a-d368cd8372e1&policy_score[min]=70&policy_score[max]=80` or accept a JSON object like `policy_score={"id": "8b409091-567f-47a7-b79a-d368cd8372e1", "range": "70..80"}` (or min/max or whatever really).

btw, ActiveRecord supports ranges; Here's an example scope that, I think, does what we want:

```rb
scope :with_policy_score, lambda { |policy, range_or_value|
  where(
    id: TestResult.latest.where(
      profile_id: policy.profiles, score: range_or_value
    ).select(:host_id)
  )
}
```

And then:
```rb
Host.with_policy_score(Policy.first, 75..85) # between 75 and 85
Host.with_policy_score(Policy.first, 75..Float::INFINITY) # greater than 75
Host.with_policy_score(Policy.first, 0..75) # less than 75
Host.with_policy_score(Policy.first, 75) # or even exactly 75
```

As you've already identified, converting from REST params and GraphQL variables into a call to that scope is the trickier part. Sorry again for not quite following during our meeting about this.

Reply:
For now I went with a minimal approach with the `current_scope` that @vkrizan suggested and it seems working. It also requires an additional param (to set the context), but without that param the query doesn't even make sense (hosts without test results)

Reply:
Well, bad news is that the `current_scope` cannot be used as other things rely on it and it kinda breaks all our tests :confused: and using the underlying `Thread.current` can leak data to a next request, so that's pretty much unsafe and it might not be a good idea to clear its value in a different context (controller) as we can't really clean it in the `filter_by_compliance_score` because we might call this method twice (less than + greater than) :confused:

Reply:
@akofink I guess the best way to get this through `scoped_search` is to pass a JSON-like object to the right side as it can hold both the ID and the interval...just need to look into this deeper if `scoped_search` likes it and how much can we force this :see_no_evil:

Reply:
OR maybe https://github.com/steveklabnik/request_store :thinking: